connectors/sources/slack.py

```diff
@@ -0,0 +1,429 @@
+                    break
+                except Exception as e:
+                    if retry >= retries:
+                        raise e
+
+                    retry += 1
+
+        return wrapped
+
+    return wrapper
+
+
+class ThrottledError(Exception):
+    """Internal exception class to indicate that request was throttled by the API"""
+
+    pass
+
+
+class SlackClient:
+    def __init__(self, configuration):
+        self.token = configuration["token"]
+        self._http_session = aiohttp.ClientSession(
+            headers=self._headers(),
+            timeout=aiohttp.ClientTimeout(total=None),
+            raise_for_status=True,
+        )
+        self._sleeps = CancellableSleeps()
+
+    def set_logger(self, logger_):
+        self._logger = logger_
+
+    def ping(self):
+        return
+
+    def close(self):
+        self._http_session.close()
+        return
+
+    async def list_channels(self, only_my_channels, cursor=None):
+        conversation_batch_limit = 200
+        self._logger.debug("Iterating over all channels")
+        if only_my_channels:
+            self._logger.debug("Will only yield channels the bot belongs to")
+        while True:
+            url = f"{BASE_URL}/{ENDPOINTS[LIST_CONVERSATIONS]}?limit={conversation_batch_limit}"
+            if cursor:
+                url = self._add_cursor(url, cursor)
+            response = await self._get_json(url)
+
+            for channel in response["channels"]:
+                if only_my_channels and not channel.get("is_member", False):
+                    continue
+                self._logger.debug(f"Yielding channel '{channel['name']}'")
+                yield channel
+            cursor = self._get_next_cursor(response)
+            if not cursor:
+                break
+
+    async def join_channel(self, channel_id):
+        url = f"{BASE_URL}/{ENDPOINTS[JOIN_CONVERSATIONS]}?channel={channel_id}"
+        return await self._get_json(url)
+
+    async def list_messages(self, channel_id, oldest, latest):
+        self._logger.info(
+            f"Fetching messages between {oldest} and {latest} from channel: '{channel_id}'"
+        )
+        message_batch_limit = 200
+        while True:
+            url = f"{BASE_URL}/{ENDPOINTS[CONVERSATION_HISTORY]}?channel={channel_id}&limit={message_batch_limit}&oldest={oldest}&latest={latest}"
+            response = await self._get_json(url)
+            for message in response.get("messages", []):
+                latest = message.get("ts")
+                if message.get("type") == "message":
+                    if message.get("reply_count", 0) > 0:
+                        async for thread_message in self.list_thread_messages(
+                            channel_id, message["ts"]
+                        ):
+                            yield thread_message
+                    else:
+                        yield message
+            if not response.get("has_more"):
+                break
```

Comment:
How does it work? do you issue same request but it gives you back different messages?

Reply:
What's changed in the request is the `latest`. See: https://github.com/elastic/connectors-python/pull/1298/files/eda90bee5e6ad4bedd200021cec9d13d7986b3ea#diff-4fc38a422f0713e89a903dce26cc081ef3afee931b5ebfdb8fab684db30c5681R130

Weirdly, slack lets you paginate by cursor, or by timestamp, but not by a combination.

Reply:
Oh wow, I see. But oldest does not change, so it's not a rolling window?

Reply:
Right - `oldest` is set by configuration (default is the last 180 days). So it needs to be a param so that configuration can have an impact, but for a given run, while paging, `oldest` stays the same, while we keep moving `latest` farther back in time until they are equal (or near enough that there are no messages between them).